src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -249,123 +256,167 @@ public async Task RefreshAsync(CancellationToken cancellationToken)
 
                     //
                     // Avoid instance state modification
-                    Dictionary<KeyValueIdentifier, ConfigurationSetting> watchedSettings = null;
+                    Dictionary<KeyValueSelector, IEnumerable<MatchConditions>> watchedFfEtags = null;
+                    Dictionary<KeyValueSelector, IEnumerable<MatchConditions>> watchedKvEtags = null;
+                    Dictionary<KeyValueIdentifier, ConfigurationSetting> watchedIndividualKvs = null;
                     List<KeyValueChange> keyValueChanges = null;
-                    List<KeyValueChange> changedKeyValuesCollection = null;
                     Dictionary<string, ConfigurationSetting> data = null;
+                    Dictionary<string, ConfigurationSetting> ffCollectionData = null;
+                    bool collectionsUpdated = false;
+                    bool ffCollectionUpdated = false;
                     bool refreshAll = false;
                     StringBuilder logInfoBuilder = new StringBuilder();
                     StringBuilder logDebugBuilder = new StringBuilder();
 
                     await ExecuteWithFailOverPolicyAsync(clients, async (client) =>
                         {
                             data = null;
-                            watchedSettings = null;
+                            ffCollectionData = null;
+                            watchedIndividualKvs = null;
+                            watchedKvEtags = null;
+                            watchedFfEtags = null;
                             keyValueChanges = new List<KeyValueChange>();
-                            changedKeyValuesCollection = null;
+                            collectionsUpdated = false;
+                            ffCollectionUpdated = false;
                             refreshAll = false;
                             Uri endpoint = _configClientManager.GetEndpointForClient(client);
                             logDebugBuilder.Clear();
                             logInfoBuilder.Clear();
 
-                            foreach (KeyValueWatcher changeWatcher in refreshableWatchers)
+                            if (!_options.RegisterAllEnabled)
                             {
-                                string watchedKey = changeWatcher.Key;
-                                string watchedLabel = changeWatcher.Label;
+                                refreshAll = await RefreshIndividualKvWatchers(
+                                    client,
+                                    keyValueChanges,
+                                    refreshableIndividualKvWatchers,
+                                    endpoint,
+                                    logDebugBuilder,
+                                    logInfoBuilder,
+                                    cancellationToken).ConfigureAwait(false);
+
+                                if (!refreshAll)
+                                {
+                                    // Get feature flag changes
```

Comment:
The code to get feature flag changes is repeated around line ~370. In order to move this logic the way it is right now into a helper, I think it would need 2 return values. Open to ideas for a better way to handle this since it alters both `watchedFfEtags` and `ffCollectionData`. I think the issue stems from `LoadSelected` basically creating 2 new dictionaries.